File: problems/arrays-hashing/encode-and-decode-strings/solution_test.go
package encode_and_decode_strings

import (
	"reflect"
	"strings"
	"testing"
)

func TestEncodeDecodeStrings(t *testing.T) {
	tests := []struct {
		name  string
		input []string
	}{
		{
			name:  "Example 1: Basic strings",
			input: []string{"hello", "world"},
		},
		{
			name:  "Example 2: Multiple strings",
			input: []string{"neet", "code", "love", "you"},
		},
		{
			name:  "Example 3: Strings with delimiter character",
			input: []string{"we", "say", ":", "yes", "%", "!"},
		},
		{
			name:  "Edge case: Empty array",
			input: []string{},
		},
		{
			name:  "Edge case: Array with empty strings",
			input: []string{"", "", ""},
		},
		{
			name:  "Edge case: Single empty string",
			input: []string{""},
		},
		{
			name:  "Complex case: Mixed empty and non-empty strings",
			input: []string{"hello", "", "world", "", "test"},
		},
		{
			name:  "Complex case: Strings with special characters",
			input: []string{"#@!", "123", "a b c", "new\nline"},
		},
		{
			name:  "Complex case: Long strings",
			input: []string{strings.Repeat("a", 1000), strings.Repeat("b", 500)},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			encoded := Encode(tt.input)
			decoded := Decode(encoded)

			if !reflect.DeepEqual(decoded, tt.input) {
				t.Errorf("Encode/Decode mismatch: got %v, want %v", decoded, tt.input)
			}
		})
	}
}

func TestEncodeEmpty(t *testing.T) {
	input := []string{}
	encoded := Encode(input)

	if encoded != "" {
		t.Errorf("Encode empty array: got %q, want empty string", encoded)
	}
}

File: problems/arrays-hashing/encode-and-decode-strings/solution.test.ts
import { describe, test, expect } from 'vitest';
import { encode, decode } from './solution';

describe('Encode and Decode Strings', () => {
  test('Example 1: Basic strings', () => {
    const input = ["hello", "world"];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Example 2: Single string', () => {
    const input = ["neet", "code", "love", "you"];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Example 3: Strings with delimiter character', () => {
    const input = ["we", "say", ":", "yes", "%", "!"];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Edge case: Empty array', () => {
    const input: string[] = [];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
    expect(encoded).toBe("");
  });

  test('Edge case: Array with empty strings', () => {
    const input = ["", "", ""];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Edge case: Single empty string', () => {
    const input = [""];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Complex case: Mixed empty and non-empty strings', () => {
    const input = ["hello", "", "world", "", "test"];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Complex case: Strings with special characters', () => {
    const input = ["#@!", "123", "a b c", "new\nline"];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });

  test('Complex case: Long strings', () => {
    const input = ["a".repeat(1000), "b".repeat(500)];
    const encoded = encode(input);
    const decoded = decode(encoded);

    expect(decoded).toEqual(input);
  });
});
